### ISSUE TITLE
Implement Differentially Private Stochastic Gradient Descent

### ISSUE THREAD

**Issue by n2cholas** [deepmind/optax#50]
Differentially Private SGD (https://cseweb.ucsd.edu/~kamalika/pubs/scs13.pdf) is an important algorithm in private machine learning. Essentially, it is SGD except you clip and add Gaussian noise to per-example gradients before averaging across the batch. I think this would be a useful addition to Optax. 

The implementation could be based on the example in the JAX repo: https://github.com/google/jax/blob/master/examples/differentially_private_sgd.py

The usage would be slightly different from other transforms, since it requires per-example gradients as inputs.  It can still be composed with other transforms as long as it is the first one in the chain. Alternatively, we can expose a stand-alone utility function that does the clipping/noise/aggregation that the user could then pass to a GradientTransform. I think the former option (making it a transform) is more convenient since this algorithm would have some state (the RNG key).

I'd be happy to work on this if it seems like a good addition.

**Comment by 8bitmp3** [comment]
Are you going to reference your paper — [Enabling Fast Differentially Private SGD via Just-in-Time Compilation and Vectorization](https://arxiv.org/pdf/2010.09063.pdf) (code: https://github.com/TheSalon/fast-dpsgd/blob/main/jaxdp.py)? 😃

**Comment by mtthss** [comment]
Hi @n2cholas, That sounds great!
I must confess I don't know much about Differentially Private SGD
but it does seem like a valuable contribution!
and since you are an expert in the topic I'd love to have you contribute it to optax!

Making it a GradientTransform sounds great! but I'll defer to your expertise on the topic
maybe we can get the author of the jax repo example to also review?

**Comment by n2cholas** [comment]
> Are you going to reference your paper — [Enabling Fast Differentially Private SGD via Just-in-Time Compilation and Vectorization](https://arxiv.org/pdf/2010.09063.pdf) (code: https://github.com/TheSalon/fast-dpsgd/blob/main/jaxdp.py)?

Thanks for bringing it up :smiley:! Our implementation was actually based on the original JAX one, so I thought it'd make more sense to reference that. Our TensorFlow 2 implementation was the complete rewrite (of TFP), which was merged into TFP actually ([PR here](https://github.com/tensorflow/privacy/pull/147)).



> Hi @n2cholas, That sounds great!
> I must confess I don't know much about Differentially Private SGD
> but it does seem like a valuable contribution!
> and since you are an expert in the topic I'd love to have you contribute it to optax!
> 
> Making it a GradientTransform sounds great! but I'll defer to your expertise on the topic
> maybe we can get the author of the jax repo example to also review?

Awesome, will do! I'll also loop in my co-authors from the fast DPSGD paper for a review once I create the PR.
